juju/juju
# storageprovisioner: fix watching of volume-backed filesystems

## Description of change

A recent change was made to the state package's filesystem and filesystem attachment watchers. The change was to filter the results based on the state of volume attachments. This doesn't work without also watching the volume attachment collection for changes, so that change has been reverted.

In its place, we introduce composite watchers in the apiserver/storageprovisioner/internal/filesystemwatcher package. These watchers compose watchers for model and machine filesystems, and volume attachments, to produce answers that the storage provisioner worker expects.

There are some incidental bug fixes, found while verifying this change:
 - the gce provider's AttachVolumes has been fixed to set DeviceLink instead of DevicePath, as in CreateVolumes
 - the storage provisioner will now react to block device changes when there are incomplete filesystem *attachments*, as well as incomplete filesystems
 - the storage provisioner removes attachments when they are Dead, not immediately after detaching the storage. This avoids an ugly "not found" error.

## QA steps

$ juju bootstrap google
$ juju deploy postgresql --storage pgdata=gce,10G
(wait)
$ juju detach-storage pgdata/0
(wait)
$ juju attach-storage postgresql/0 pgdata/0
(storage should show up as attached)

## Documentation changes

None.

## Bug reference

None.
